javaslang/generator/Generator.scala

```diff
@@ -2408,49 +2394,37 @@ def generateMainClasses(): Unit = {
           }
 
           @SuppressWarnings("unchecked")
+          static <T> Object[] asArray(Iterable<? extends T> iterable) {
+              if (iterable instanceof Collection<?>) {
+                  final Collection<? extends T> collection = (Collection<? extends T>) iterable;
+                  return collection.toArray();
+              } else {
+                  final Tuple2<Iterable<? extends T>, Integer> iterableAndSize = Collections.withSize(iterable);
+                  return asArray(iterableAndSize._1.iterator(), iterableAndSize._2);
+              }
+          }
+
+          @SuppressWarnings("unchecked")
           static <T> T asPrimitives(Class<?> primitiveClass, Iterable<?> values) {
               final Object[] array = Array.ofAll(values).toJavaArray();
-              assert (array.length == 0) || (primitiveClass == primitiveType(array[0])) && !primitiveClass.isArray();
+              assert (array.length == 0) || !primitiveClass.isArray();
               final ArrayType<T> type = of((Class<T>) primitiveClass);
               final Object results = type.newInstance(array.length);
               for (int i = 0; i < array.length; i++) {
-                  type.setAt(results, i, array[i]);
+                  type.setAt(results, i, (T) array[i]);
               }
               return (T) results;
           }
 
-          static <T> Class<?> primitiveType(T element) {
-              final Class<?> wrapper = (element == null) ? Object.class : element.getClass();
-              if (wrapper == Boolean.class) {
-                  return boolean.class;
-              } else if (wrapper == Byte.class) {
-                  return byte.class;
-              } else if (wrapper == Character.class) {
-                  return char.class;
-              } else if (wrapper == Double.class) {
-                  return double.class;
-              } else if (wrapper == Float.class) {
-                  return float.class;
-              } else if (wrapper == Integer.class) {
-                  return int.class;
-              } else if (wrapper == Long.class) {
-                  return long.class;
-              } else if (wrapper == Short.class) {
-                  return short.class;
-              } else {
-                  return wrapper;
-              }
-          }
-
-          ${types.keys.toSeq.sorted.gen(arrayType =>
+          ${types.keys.toSeq.gen(arrayType =>
```

Comment:
managed to get rid of this by relying on the runtime to throw a `ClassCastException`, instead of checking every insertion for correctness (i.e. a `null` or `String` value in a primitive array)